nexus/src/app/background/tasks/tuf_artifact_replication.rs

```diff
@@ -0,0 +1,447 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! TUF Repo Depot: Artifact replication across sleds (RFD 424)
+//!
+//! `Nexus::updates_put_repository` accepts a TUF repository, which Nexus
+//! unpacks, verifies, and reasons about the artifacts in. This uses temporary
+//! storage within the Nexus zone, so the update artifacts have to go somewhere.
+//! We've settled for now on "everywhere": a copy of each artifact is stored on
+//! each sled's M.2 devices.
+//!
+//! This background task is responsible for getting locally-stored artifacts
+//! onto sleds, and ensuring all sleds have copies of all artifacts.
+//! `Nexus::updates_put_repository` sends the [`ArtifactsWithPlan`] object to
+//! this task via an [`mpsc`] channel and activates it.
+//!
+//! During each activation:
+//!
+//! 1. The task moves `ArtifactsWithPlan` objects of the `mpsc` channel and into
+//!    a `Vec`.
+//! 2. The task queries the list of artifacts stored on each sled, and compares
+//!    it to the list of artifacts in CockroachDB. Sled artifact storage is
+//!    content-addressed by SHA-256 checksum. Errors are logged but otherwise
+//!    ignored (unless no sleds respond); the task proceeds as if that sled
+//!    has no artifacts. (This means that the task will always be trying to
+//!    replicate artifacts to that sled until it comes back or is pulled out
+//!    of service.)
+//! 3. The task compares the list of artifacts from the database to the list of
+//!    artifacts available locally.
+//! 4. If all the artifacts belonging to an `ArtifactsWithPlan` object have
+//!    been replicated to at least `MIN_SLED_REPLICATION` sleds, the task drops
+//!    the object from its `Vec` (thus cleaning up the local storage of those
+//!    files).
+//! 5. The task generates a list of requests that need to be sent:
+//!    - PUT each locally-stored artifact not present on any sleds to a random
+//!      sled.
+//!    - For each partially-replicated artifact, choose a sled that is missing
+//!      the artifact, and tell it (via `artifact_copy_from_depot`) to fetch the
+//!      artifact from a random sled that has it.
+//!    - DELETE all artifacts no longer tracked in CockroachDB from all sleds
+//!      that have that artifact.
```

Comment:
Not yet. This is because we don't yet have an endpoint to delete a repository, and I don't think it's safe to add one until we know what artifacts are currently being "used" by the control plane to prevent deletion of those repositories. (And I think that needs to be outside the scope of this particular PR.)

In the limit I think we will have some idea of a current system version and a target system version, and Nexus will need to protect those versions and all the intermediate versions from deletion so that it can perform updates using the available artifacts.

Reply:
Although, I probably could figure out how to add an arbitrary artifact to the simulated sled agent and ensure it's deleted...